engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java

```diff
@@ -1976,92 +1965,244 @@ public void storageMigration(final String vmUuid, final StoragePool destPool) {
     private void orchestrateStorageMigration(final String vmUuid, final StoragePool destPool) {
         final VMInstanceVO vm = _vmDao.findByUuid(vmUuid);
 
+        preStorageMigrationStateCheck(destPool, vm);
+
+        try {
+            if(s_logger.isDebugEnabled()) {
+                s_logger.debug(
+                        String.format("Offline migration of %s vm %s with volumes",
+                                vm.getHypervisorType().toString(),
+                                vm.getInstanceName()));
+            }
+
+            migrateThroughHypervisorOrStorage(destPool, vm);
+
+        } catch (ConcurrentOperationException
+                | InsufficientCapacityException // possibly InsufficientVirtualNetworkCapacityException or InsufficientAddressCapacityException
+                | StorageUnavailableException e) {
+            s_logger.debug("Failed to migration: " + e.toString());
+            throw new CloudRuntimeException("Failed to migration: " + e.toString());
+        } finally {
+            try {
+                stateTransitTo(vm, Event.AgentReportStopped, null);
+            } catch (final NoTransitionException e) {
+                s_logger.debug("Failed to change vm state: " + e.toString());
+                throw new CloudRuntimeException("Failed to change vm state: " + e.toString());
+            }
+        }
+    }
+
+    private Answer[] attemptHypervisorMigration(StoragePool destPool, VMInstanceVO vm) {
+        boolean migrationResult = false;
+        final HypervisorGuru hvGuru = _hvGuruMgr.getGuru(vm.getHypervisorType());
+        // OfflineVmwareMigration: in case of vmware call vcenter to do it for us.
+        // OfflineVmwareMigration: should we check the proximity of source and destination
+        // OfflineVmwareMigration: if we are in the same cluster/datacentre/pool or whatever?
+        // OfflineVmwareMigration: we are checking on success to optionally delete an old vm if we are not
+        List<Command> commandsToSend = hvGuru.finalizeMigrate(vm, destPool);
+
+        Long hostId = vm.getHostId();
+        // OfflineVmwareMigration: probaby this is null when vm is stopped
+        if(hostId == null) {
+            hostId = vm.getLastHostId();
+            if (s_logger.isDebugEnabled()) {
+                s_logger.debug(String.format("host id is null, using last host id %d", hostId) );
+            }
+        }
+
+        if(CollectionUtils.isNotEmpty(commandsToSend)) {
+            Commands commandsContainer = new Commands(commandsToSend.get(0));
```

Comment:
are you only sending the first command?

Reply:
good point, yes we are actually but the framework dictates. we might need to change this on a refactor though.

Reply:
Agreed. How about this instead?

````
Commands commandsContainer = new Commands(Command.OnError.Stop);
commandsContainer.addCommands(commandsToSend);
````